### PR TITLE
feat(triage): Phase 1 mechanical issue triage + incremental scan

### LINKED ISSUE #2259: feat(triage): Phase 1 - mechanical issue triage + incremental scan state (epic #1799)
**Parent epic:** #1799
**Blocks:** Phase 2 (AI triage), Phase 3 (recommendations)

## Goal

Phase 1 of the AI issue-triage pipeline (#1799): the mechanical, scriptable triage layer (no LLM) plus the incremental-scan state foundation that later phases build on.

## Scope

- **Stale detection**: issues with no activity in N days -> label stale, suggest close.
- **Duplicate matching**: compare titles + bodies for similarity -> suggest merge.
- **Label hygiene**: issues missing required labels (priority, area, agent) -> suggest labels.
- **State consistency**: issues in wrong state (Doing with no active worker, Planning with no description) -> flag.
- **Linked PR status**: issues with merged/closed linked PRs -> auto-advance state.
- **Incremental scan state**: persist a "last triage run" marker so subsequent runs only scan issues modified since the last run. This is the shared foundation the spec calls for ("only scan issues modified since last triage run") and every later phase reuses it.

## Acceptance criteria

- [ ] Script performs all five mechanical checks above and emits structured suggestions (no auto-execute).
- [ ] Incremental scan: a persisted timestamp/state filter limits a run to issues changed since the previous run; full-scan is opt-in.
- [ ] Read-only by default: triage suggests, human approves.
- [ ] Parallel-safe: multiple scanners can process disjoint issue sets.
- [ ] Tests: positive, negative, and edge cases for each check; mock GitHub I/O.

## Out of scope

- AI/LLM classification (Phase 2, #1799).
- Recommendation report + batch apply (Phase 3, #1799).

## Source

Decomposed from epic #1799 (ClawSweeper pattern). See epic for full context and references.

### PR DESCRIPTION
**Pull request by rjmurillo**
## Summary

Phase 1 of the ClawSweeper issue-triage pipeline (epic #1799): the mechanical, no-LLM triage layer plus the incremental-scan foundation later phases build on. Extends the existing `scripts/issue_triage.py` (which shipped stale + priority/area label checks) rather than creating a parallel script.

Read-only by design: emits structured JSON or human-readable suggestions, never mutates GitHub state.

## New mechanical checks

| Check | Function | Rule |
|-------|----------|------|
| Missing agent label | `has_agent_label` | no `agent-*` label |
| State consistency | `check_state_consistency` | `Doing` with no assignee, or `Planning` with empty body |
| Linked PR status | `detect_linked_pr_status` | a linked PR merged/closed so the state label can advance |
| Duplicate candidates | `detect_duplicates` | pairwise title Jaccard similarity over normalized tokens (conventional-commit prefix stripped), threshold via `--dup-threshold` (default 0.7) |

## Incremental scan

- `--state-file` persists `{"last_run": ISO8601}`. The next run restricts the gh query to `updated:>last_run`.
- `--full-scan` ignores the state file; `--since` overrides it.
- State is saved **only after a successful scan**, so a failed run does not advance the watermark and silently skip issues.
- This keeps repeated runs cheap and lets multiple scanners process disjoint issue sets in parallel.

## Design notes

- Linked-PR detection is the only per-issue API call and is **opt-in** (`--check-linked-prs`), keeping the default scan fast and parallel-safe.
- `IssueRecord` gains `body`, `assignees`, `linked_prs` (all defaulted, backward compatible).
- `classify()` retained as the stable 3-rule entry point; `build_report` layers the new rules on top (DRY).
- `main()` decomposed into `_validate_args` / `_obtain_raw_issues` / `_parse_records` / `_emit` (cyclomatic complexity 12 to <=10).

## Out of scope (tracked separately)

- AI/LLM classification: Phase 2, #2260.
- Recommendation report + batch apply: Phase 3, #2261.

## Tests

76 tests (was 46), mypy clean, ruff clean. New classes: agent-label, state consistency, linked-PR status, duplicate detection (incl. threshold boundary and conventional-prefix handling), incremental scan (load/save roundtrip, `--since` precedence, `--full-scan`, state persisted after run). Two baseline tests updated for the new agent-label rule and the 7-section report, each with a comment citing the behavior change. Verified live read-only against this repo.

Fixes #2259

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/issue_triage.py [hunk 1]
```diff
@@ -225,42 +471,113 @@ def fetch_open_issues(owner: str, repo: str, *, limit: int) -> list[dict]:
...
+    state_path = Path(path)
+    if not state_path.is_file():
+        return None
+    try:
+        data = json.loads(state_path.read_text(encoding="utf-8"))
+    except (OSError, json.JSONDecodeError, ValueError):
+        return None
+    last_run = data.get("last_run") if isinstance(data, dict) else None
+    if isinstance(last_run, str) and last_run:
+        return last_run
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/issue_triage.py:548]: Fixed in 321b6855: load_scan_state now returns last_run only when it matches ISO_TIMESTAMP_PATTERN. Invalid state is treated as missing state, with regression coverage in tests/test_issue_triage.py.

File: scripts/issue_triage.py [hunk 2]
```diff
@@ -295,62 +626,162 @@ def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
...
+def _validate_args(args: argparse.Namespace) -> int | None:
+    """Return an error exit code for invalid arguments, or None when valid."""
 
     if args.stale_days < 0:
         print("--stale-days must be >= 0", file=sys.stderr)
         return 2
+    if not 0.0 <= args.dup_threshold <= 1.0:
+        print("--dup-threshold must be between 0.0 and 1.0", file=sys.stderr)
+        return 2
+    return None
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/issue_triage.py:773]: Fixed in 321b6855: _validate_args now rejects invalid --since and out-of-range --limit as config errors. Limit 0 remains valid by design for empty discovery runs, and tests cover that path.

File: scripts/issue_triage.py [hunk 3]
```diff
@@ -225,42 +471,113 @@ def fetch_open_issues(owner: str, repo: str, *, limit: int) -> list[dict]:
...
+    if not state_path.is_file():
+        return None
+    try:
+        data = json.loads(state_path.read_text(encoding="utf-8"))
+    except (OSError, json.JSONDecodeError, ValueError):
+        return None
+    last_run = data.get("last_run") if isinstance(data, dict) else None
+    if isinstance(last_run, str) and last_run:
+        return last_run
+    return None
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/issue_triage.py:549]: Fixed in 321b6855: corrupted last_run values no longer flow into the gh search filter. Added an invalid timestamp state-file test.

File: scripts/issue_triage.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/issue_triage.py:773]: Fixed in 321b6855: --since validation now uses the same ISO timestamp pattern and exits 2 on bad input.

File: scripts/issue_triage.py [hunk 4]
```diff
@@ -225,42 +471,113 @@ def fetch_open_issues(owner: str, repo: str, *, limit: int) -> list[dict]:
...
+    cmd = [
+        "gh", "api",
+        f"repos/{owner}/{repo}/issues/{number}/timeline",
+        "--paginate",
+    ]
+    try:
+        raw = _run_gh(cmd)
+        events = json.loads(raw or "[]")
+    except (RuntimeError, json.JSONDecodeError, ValueError, TypeError):
+        return ()
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/issue_triage.py]: Fixed in 321b6855: linked-PR timeline failures now fail the scan with exit 3, so the incremental watermark is not advanced after incomplete enrichment.

File: scripts/issue_triage.py [hunk 5]
```diff
@@ -295,62 +626,162 @@ def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
...
         stale_days=args.stale_days,
+        dup_threshold=args.dup_threshold,
+        scanned_since=since,
     )
+    _emit(report, args.format)
+
+    # Persist the run timestamp only after a successful scan, so a failed run
+    # does not advance the watermark and skip issues next time.
+    if args.state_file and not args.input:
+        save_scan_state(args.state_file, report.timestamp)
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/issue_triage.py]: Fixed in 321b6855: state-file write failures are caught and returned as exit 2 with a clear error instead of an uncaught traceback.

File: scripts/issue_triage.py [hunk 6]
```diff
@@ -357,54 +711,147 @@ def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
...
+    from dataclasses import replace
+
+    enriched: list[IssueRecord] = []
+    for issue in issues:
+        try:
+            linked = fetch_linked_prs(owner, repo, issue.number)
+        except LinkedPrFetchError as err:
+            raise _InputError(
+                3, f"failed to fetch linked PRs for issue #{issue.number}: {err}"
+            ) from err
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/issue_triage.py:744]: Fixed in 80ca21ae: fetch_linked_prs now raises only the root cause, and _enrich_linked_prs adds the issue context once.

File: scripts/issue_triage.py [hunk 7]
```diff
@@ -418,12 +861,20 @@ def main(argv: list[str] | None = None) -> int:
...
+        dup_threshold=args.dup_threshold,
+        scanned_since=since,
     )
+    _emit(report, args.format)
+
+    # Persist the run timestamp only after a successful scan, so a failed run
+    # does not advance the watermark and skip issues next time.
+    if args.state_file and not args.input:
+        try:
+            save_scan_state(args.state_file, report.timestamp)
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/issue_triage.py]: Fixed in 8380b7d6: the incremental state now persists a scan-start timestamp captured before fetching, so updates during the scan are picked up on the next run.

File: scripts/issue_triage.py [hunk 8]
```diff
@@ -119,11 +211,16 @@ def parse_issue_record(raw: dict) -> IssueRecord:
...
+
+    body = raw.get("body")
     return IssueRecord(
         number=number,
         title=title,
         updated_at=updated_at,
         labels=tuple(labels),
+        body=str(body) if body is not None else "",
+        assignees=_parse_assignees(raw.get("assignees")),
+        linked_prs=_parse_linked_prs(raw.get("linkedPrs")),
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/issue_triage.py]: Fixed in 8380b7d6: parse_issue_record accepts both linkedPrs and linked_prs input keys, with regression coverage for snake_case input.

File: scripts/issue_triage.py [hunk 9]
```diff
@@ -151,46 +248,151 @@ def has_area_label(issue: IssueRecord) -> bool:
...
+    """Jaccard similarity of two token sets: |A and B| / |A or B|.
+
+    Returns 0.0 when both sets are empty (no signal to compare).
+    """
+
+    if not a and not b:
+        return 0.0
+    union = a | b
+    if not union:
+        return 0.0
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/issue_triage.py]: Fixed in 8380b7d6: removed the unreachable union emptiness branch from jaccard_similarity.
